mdagent/tools/base_tools/analysis_tools/distance_tools.py

```diff
@@ -0,0 +1,543 @@
+            mode = "CA"
+        if analysis not in ["all", "not all"]:
+            print(
+                "Incorrect analysis, must be either 'all' or 'not all'."
+                " Defaulting to all"
+            )
+            system_message += "'analysis', must be either 'all' or 'not all'.\
+                  Tool defaulted to 'all' \n"
+            analysis = "all"
+        keys = input.keys()
+        for key in keys:
+            if key not in [
+                "trajectory_fileid",
+                "topology_fileid",
+                "analysis",
+                "mode",
+                "selection1",
+                "selection2",
+            ]:
+                system_message += (
+                    f"{key} is not part of admitted tool inputs\n."
+                    "Ignoring it during the analysis"
+                )
+
+        if error == "":
+            error = None
+        return {
+            "trajectory_fileid": trajectory_id,
+            "topology_fileid": topology_id,
+            "selection1": selection1,
+            "selection2": selection2,
+            "analysis": analysis,
+            "mode": mode,
+            "error": error,
+            "system_message": system_message,
+        }
+
+    def get_values(self, input):
+        traj_id = input.get("trajectory_fileid")
+        top_id = input.get("topology_fileid")
+        sel1 = input.get("selection1")
+        sel2 = input.get("selection2")
+        analysis = input.get("analysis")
+        mode = input.get("mode")
+        error = input.get("error")
+        syst_mes = input.get("system_message")
+
+        return traj_id, top_id, sel1, sel2, analysis, mode, error, syst_mes
+
+
+class ContactsTool(BaseTool):
+    name = "ContactsTool"
+    description = (
+        "Tool for computing the distance between pairs of residues in a trajectory. "
+        "If distance is under the cutoff is considered a contact. The output is a "
+        "matrix plot where each contact between residues is represented by a dot."
+    )
+    input_schema = ContactSchema
+    path_registry: Optional[PathRegistry]
+
+    def __init__(self, path_registry: Optional[PathRegistry] = None):
+        super().__init__()
+        self.path_registry = path_registry
+
+    def _run(self, **input):
+        try:
+            input = self.validate_input(**input)
+        except ValueError as e:
+            return f"Failed. Error using the Contacts Tool: {str(e)}"
+        (
+            trajectory_id,
+            topology_id,
+            selection,
+            cutoff,
+            error,
+            system_message,
+        ) = self.get_values(input)
+
+        if error:
+            return f"Failed. Error with the tool inputs: {error} "
+
+        path_to_traj = self.path_registry.get_mapped_path(trajectory_id)
+        path_to_top = self.path_registry.get_mapped_path(topology_id)
+
+        try:
+            traj = md.load(path_to_traj, top=path_to_top)
```

Comment:
do you want to use "load_single_traj" function from data_handling.py in utils folder? it'll save a lot of lines here. Feel free to adjust that function if it doesn't suit your need exactly.